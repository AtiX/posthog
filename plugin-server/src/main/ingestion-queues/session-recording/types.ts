// This is the incoming message from Kafka

import { RRWebEvent } from '../../../types'

export type RawHeatmapEvent = {
    /**
     * session id lets us offer example recordings on high traffic parts of the page,
     * and could let us offer more advanced filtering of heatmap data
     * we will break the relationship between particular sessions and clicks in aggregating this data
     * it should always be treated as an exemplar and not as concrete values
     */
    $session_id: string
    $viewport_width: number
    $viewport_height: number
    $pointer_target_fixed: boolean
}

export type HeatmapEvent = RawHeatmapEvent & {
    // x is the x with resolution applied, the resolution converts high fidelity mouse positions into an NxN grid
    x: number
    // y is the y with resolution applied, the resolution converts high fidelity mouse positions into an NxN grid
    y: number
    scale_factor: 16 // in the future we may support other values
    timestamp: string
}

export type IncomingHeatmapEventMessage = RawHeatmapEvent & {
    metadata: {
        topic: string
        partition: number
        timestamp: number
    }
    team_id: number
    $pointer_x: number
    $pointer_y: number
}

export type IncomingRecordingMessage = {
    metadata: {
        topic: string
        partition: number
        rawSize: number
        lowOffset: number
        highOffset: number
        timestamp: number
        consoleLogIngestionEnabled?: boolean
    }

    team_id: number
    distinct_id: string
    session_id: string
    eventsByWindowId: Record<string, RRWebEvent[]>
    eventsRange: {
        start: number
        end: number
    }
    snapshot_source: string | null
}

export type PersistedRecordingMessage = {
    window_id?: string
    data: any
}
