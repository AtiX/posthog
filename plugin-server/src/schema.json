{
    "$ref": "#/definitions/SummarizedSessionRecordingEvent",
    "$schema": "http://json-schema.org/draft-07/schema#",
    "definitions": {
        "SummarizedSessionRecordingEvent": {
            "additionalProperties": false,
            "description": "written to kafka topic: clickhouse_session_replay_events and from there to clickhouse table: session_replay_events",
            "properties": {
                "active_milliseconds": {
                    "type": "integer"
                },
                "click_count": {
                    "type": "integer"
                },
                "console_error_count": {
                    "type": "integer"
                },
                "console_log_count": {
                    "type": "integer"
                },
                "console_warn_count": {
                    "type": "integer"
                },
                "distinct_id": {
                    "type": "string"
                },
                "first_timestamp": {
                    "type": "string"
                },
                "first_url": {
                    "type": ["string", "null"]
                },
                "keypress_count": {
                    "type": "integer"
                },
                "last_timestamp": {
                    "type": "string"
                },
                "mouse_activity_count": {
                    "type": "integer"
                },
                "session_id": {
                    "type": "string"
                },
                "size": {
                    "type": "integer"
                },
                "team_id": {
                    "type": "number"
                },
                "uuid": {
                    "type": "string"
                }
            },
            "required": [
                "uuid",
                "first_timestamp",
                "last_timestamp",
                "team_id",
                "distinct_id",
                "session_id",
                "click_count",
                "keypress_count",
                "mouse_activity_count",
                "active_milliseconds",
                "console_log_count",
                "console_warn_count",
                "console_error_count",
                "size"
            ],
            "type": "object"
        }
    }
}
