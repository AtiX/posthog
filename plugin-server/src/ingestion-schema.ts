/**
* Types added to this file will be processed into the Plugin Server schema.json file and be available for validation using `ajv`
*/

/**
 *
 * written to kafka topic: clickhouse_session_replay_events
 * and from there to clickhouse table: session_replay_events
 */
export interface SummarizedSessionRecordingEvent {
    uuid: string
    first_timestamp: string
    last_timestamp: string
    team_id: number
    distinct_id: string
    session_id: string
    first_url: string | undefined
    click_count: number
    keypress_count: number
    mouse_activity_count: number
    active_milliseconds: number
    console_log_count: number
    console_warn_count: number
    console_error_count: number
    size: number
}
