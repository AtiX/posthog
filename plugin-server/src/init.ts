import * as Pyroscope from '@pyroscope/nodejs'
import { hostname } from 'os'

import { initSentry } from './sentry'
import { PluginsServerConfig } from './types'
import { status } from './utils/status'

// Code that runs on app start, in both the main and worker threads
export function initApp(config: PluginsServerConfig): void {
    initSentry(config)
    initPyroscope(config)
}

export function initPyroscope(config: PluginsServerConfig): void {
    if (config.PYROSCOPE_ADDRESS && config.PYROSCOPE_TOKEN) {
        let tags: Record<string, any> = {}
        try {
            tags = JSON.parse(config.PYROSCOPE_EXTRA_TAGS)
        } catch (error) {
            status.warn('ðŸ’¥', 'Invalid PYROSCOPE_EXTRA_TAGS format:', error)
        } finally {
            tags.pod_name = hostname()
        }
        Pyroscope.init({
            serverAddress: config.PYROSCOPE_ADDRESS,
            authToken: config.PYROSCOPE_TOKEN,
            appName: 'plugin-server-' + config.PLUGIN_SERVER_MODE,
            tags: tags,
        })
    }
}
