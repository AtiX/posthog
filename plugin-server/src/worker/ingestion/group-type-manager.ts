import LRU from 'lru-cache'

import { GroupTypeIndex, GroupTypeToColumnIndex, Team, TeamId } from '../../types'
import { DB } from '../../utils/db/db'
import { timeoutGuard } from '../../utils/db/utils'
import { posthog } from '../../utils/posthog'
import { TeamManager } from './team-manager'

export class GroupTypeManager {
    db: DB
    teamManager: TeamManager
    groupTypesCache: LRU<number, GroupTypeToColumnIndex>
    instanceSiteUrl: string

    constructor(db: DB, teamManager: TeamManager, instanceSiteUrl?: string | null) {
        this.db = db
        this.teamManager = teamManager
        this.groupTypesCache = new LRU({
            max: 10_000,
            // 30 seconds
            maxAge: 30_000,
            // being explicit about the fact that we want to update
            // the team cache every 2min, irrespective of the last access
            updateAgeOnGet: false,
        })
        this.instanceSiteUrl = instanceSiteUrl || 'unknown'
    }

    public async fetchGroupTypes(teamId: TeamId): Promise<GroupTypeToColumnIndex> {
        const cachedGroupTypes = this.groupTypesCache.get(teamId)
        if (cachedGroupTypes) {
            return cachedGroupTypes
        }

        const timeout = timeoutGuard(`Still running "fetchGroupTypes". Timeout warning after 30 sec!`)
        try {
            const teamGroupTypes: GroupTypeToColumnIndex = await this.db.fetchGroupTypes(teamId)
            this.groupTypesCache.set(teamId, teamGroupTypes)
            return teamGroupTypes
        } finally {
            clearTimeout(timeout)
        }
    }

    public async fetchGroupTypeIndex(teamId: TeamId, groupType: string): Promise<GroupTypeIndex | null> {
        const groupTypes = await this.fetchGroupTypes(teamId)

        if (groupType in groupTypes) {
            return groupTypes[groupType]
        } else {
            const [groupTypeIndex, isInsert] = await this.db.insertGroupType(
                teamId,
                groupType,
                Object.keys(groupTypes).length
            )
            if (groupTypeIndex !== null) {
                this.groupTypesCache.del(teamId)
            }

            if (isInsert && groupTypeIndex !== null) {
                await this.captureGroupTypeInsert(teamId, groupType, groupTypeIndex)
            }
            return groupTypeIndex
        }
    }

    private async captureGroupTypeInsert(teamId: TeamId, groupType: string, groupTypeIndex: GroupTypeIndex) {
        const team: Team | null = await this.teamManager.fetchTeam(teamId)

        if (!team) {
            return
        }

        posthog.capture({
            distinctId: 'plugin-server',
            event: 'group type ingested',
            properties: {
                team: team.uuid,
                groupType,
                groupTypeIndex,
            },
            groups: {
                project: team.uuid,
                organization: team.organization_id,
                instance: this.instanceSiteUrl,
            },
        })
    }
}
