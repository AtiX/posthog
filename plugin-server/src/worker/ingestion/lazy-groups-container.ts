import { DB, GroupId, GroupsData } from '../../utils/db/db'

export class LazyGroupsContainer {
    teamId: number
    groupIds: GroupId[]

    loaded: boolean

    private db: DB
    private promise: Promise<GroupsData> | null

    constructor(teamId: number, groupIds: GroupId[], db: DB) {
        this.teamId = teamId
        this.groupIds = groupIds
        this.db = db

        this.promise = null
        this.loaded = false
    }

    async get(): Promise<GroupsData> {
        if (!this.promise) {
            this.promise = this.db.getGroupsData(this.teamId, this.groupIds).then((groupsData) => {
                if (groupsData) {
                    this.loaded = true
                }
                return groupsData
            })
        }
        return this.promise
    }
}
