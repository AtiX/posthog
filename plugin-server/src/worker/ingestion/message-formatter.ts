import { format } from 'util'

import { PostIngestionEvent, Team } from '../../types'
import { getPropertyValueByPath, stringify } from '../../utils/utils'

// Sync with .../api/person.py and .../lib/constants.tsx
const PERSON_DEFAULT_DISPLAY_NAME_PROPERTIES = ['email', 'Email', 'name', 'Name', 'username', 'Username', 'UserName']

const TOKENS_REGEX_BRACKETS_EXCLUDED = /(?<=(?<!\\){{)(.*?)(?=(?<!\\)}})/g
const TOKENS_REGEX_BRACKETS_INCLUDED = /(?<!\\){{(.*?)(?<!\\)\}}/g

export type MessageFormatterOptions = {
    event: PostIngestionEvent
    team: Team
    siteUrl: string
    sourcePath: string
    sourceName: string
}

// This formatter is "simpler" in that it only supports parsing of a few tokens
export class MessageFormatter {
    private projectUrl: string
    private personLink: string
    private eventLink: string
    private sourceLink: string

    constructor(private options: MessageFormatterOptions) {
        this.projectUrl = `${options.siteUrl}/project/${options.team.id}`

        this.personLink = `${this.projectUrl}/person/${encodeURIComponent(options.event.distinctId)}`
        this.eventLink = `${this.projectUrl}/events/${encodeURIComponent(options.event.eventUuid)}/${encodeURIComponent(
            options.event.timestamp
        )}`

        this.sourceLink = `${this.projectUrl}${options.sourcePath}`
    }

    format(template: string): string {
        const message = this.getFormattedMessage(template)
        return message
    }

    formatSafely(template: unknown): unknown {
        if (typeof template === 'string') {
            return this.getFormattedMessage(template)
        }

        return template
    }

    private getFormattedMessage(template: string): string {
        try {
            const [tokens, tokenizedMessage] = this.getTokens(template)
            const values: string[] = []

            for (const token of tokens) {
                const tokenParts = token.split('.') || []

                const value = this.getValueOfToken(tokenParts)

                console.log(tokenParts, value, tokenizedMessage)
                values.push(value)
            }
            return format(tokenizedMessage, ...values)
        } catch (error) {
            return `âš  Error: There are one or more formatting errors in the message template for source "${this.options.sourceName} - ${this.sourceLink}".`
        }
    }

    private escapeMarkdown(text: string): string {
        const markdownChars: string[] = ['\\', '`', '*', '_', '{', '}', '[', ']', '(', ')', '!']
        const lineStartChars: string[] = ['#', '-', '+']

        let escapedText = ''
        let isNewLine = true

        for (const char of text) {
            if (isNewLine && lineStartChars.includes(char)) {
                escapedText += '\\' + char
            } else if (!isNewLine && markdownChars.includes(char)) {
                escapedText += '\\' + char
            } else {
                escapedText += char
            }

            isNewLine = char === '\n' || char === '\r'
        }

        return escapedText
    }

    private webhookEscape(text: string): string {
        return this.escapeMarkdown(stringify(text))
    }

    private getPersonDetails(): string {
        // Sync the logic below with the frontend `asDisplay`
        const personDisplayNameProperties =
            this.options.team.person_display_name_properties ?? PERSON_DEFAULT_DISPLAY_NAME_PROPERTIES
        const customPropertyKey = personDisplayNameProperties.find((x) => this.options.event.person_properties?.[x])
        const propertyIdentifier = customPropertyKey
            ? this.options.event.person_properties[customPropertyKey]
            : undefined

        const customIdentifier: string =
            typeof propertyIdentifier !== 'string' ? JSON.stringify(propertyIdentifier) : propertyIdentifier

        const display: string | undefined = (customIdentifier || this.options.event.distinctId)?.trim()

        return display
    }

    private getGroupLink(groupIndex: number, groupKey: string): string {
        return `${this.projectUrl}/groups/${groupIndex}/${encodeURIComponent(groupKey)}`
    }

    private getTokens(template: string): [string[], string] {
        // This finds property value tokens, basically any string contained in square brackets
        // Examples: "[foo]" is matched in "bar [foo]", "[action.name]" is matched in "action [action.name]"
        // The backslash is used as an escape character - "\[foo\]" is not matched, allowing square brackets in messages
        const matchedTokens = template.match(TOKENS_REGEX_BRACKETS_EXCLUDED) || []
        // Replace the tokens with placeholders, and unescape leftover brackets
        const tokenizedMessage = template.replace(TOKENS_REGEX_BRACKETS_INCLUDED, '%s').replace(/\\({{|}})/g, '$1')
        return [matchedTokens, tokenizedMessage]
    }

    private getValueOfToken(tokenParts: string[]): string {
        let text = ''

        if (tokenParts[0] === 'user') {
            // [user.name] and [user.foo] are DEPRECATED as they had odd mechanics
            // [person] OR [event.properties.bar] should be used instead
            if (tokenParts[1] === 'name') {
                text = this.getPersonDetails()
            } else {
                const propertyName = `$${tokenParts[1]}`
                const property = this.options.event.properties?.[propertyName]
                text = this.webhookEscape(property)
            }
        } else if (tokenParts[0] === 'person') {
            if (tokenParts.length === 1) {
                text = stringify({
                    id: this.options.event.person_id,
                    properties: this.options.event.person_properties,
                    link: this.personLink,
                })
            } else if (tokenParts[1] === 'link') {
                text = this.webhookEscape(this.personLink)
            } else if (tokenParts[1] === 'properties') {
                if (tokenParts.length == 2) {
                    text = stringify(this.options.event.person_properties)
                } else {
                    const property = this.options.event.person_properties
                        ? getPropertyValueByPath(this.options.event.person_properties, tokenParts.slice(2))
                        : undefined
                    text = this.webhookEscape(property)
                }
            }
        } else if (tokenParts[0] === 'action' || tokenParts[0] === 'source') {
            if (tokenParts[1] === 'name') {
                text = this.options.sourceName
            } else if (tokenParts[1] === 'link') {
                text = this.webhookEscape(this.sourceLink)
            }
        } else if (tokenParts[0] === 'event') {
            if (tokenParts.length === 1) {
                // TODO: Standardise this format
                text = stringify({
                    uuid: this.options.event.eventUuid,
                    event: this.options.event.event,
                    timestamp: this.options.event.timestamp,
                    distinct_id: this.options.event.distinctId,
                    link: this.eventLink,
                    properties: this.options.event.properties,
                })
            } else if (tokenParts[1] === 'link') {
                text = this.webhookEscape(this.eventLink)
            } else if (tokenParts[1] === 'uuid') {
                text = this.webhookEscape(this.options.event.eventUuid)
            } else if (tokenParts[1] === 'name') {
                // deprecated
                text = this.webhookEscape(this.options.event.event)
            } else if (tokenParts[1] === 'event') {
                text = this.webhookEscape(this.options.event.event)
            } else if (tokenParts[1] === 'timestamp') {
                text = this.webhookEscape(this.options.event.timestamp)
            } else if (tokenParts[1] === 'distinct_id') {
                text = this.webhookEscape(this.options.event.distinctId)
            } else if (tokenParts[1] === 'properties') {
                if (tokenParts.length === 2) {
                    text = stringify(this.options.event.properties)
                } else {
                    const property = this.options.event.properties
                        ? getPropertyValueByPath(this.options.event.properties, tokenParts.slice(2))
                        : undefined
                    text = this.webhookEscape(property)
                }
            }
        } else if (tokenParts[0] === 'groups') {
            if (tokenParts.length === 1) {
                text = stringify(this.options.event.groups)
            } else if (tokenParts.length > 1) {
                const relatedGroup = this.options.event.groups?.[tokenParts[1]]

                if (!relatedGroup) {
                    // What to return if no matching group?
                    return `(event without group '${tokenParts[1]}')`
                }

                if (tokenParts.length === 2) {
                    // Return the group name
                    // NOTE: group type is not correct...
                    return stringify(relatedGroup)
                } else if (tokenParts[2] === 'key') {
                    text = this.webhookEscape(relatedGroup.key)
                } else if (tokenParts[2] === 'link') {
                    text = this.webhookEscape(this.getGroupLink(relatedGroup.index, relatedGroup.key))
                } else if (tokenParts[2] === 'properties') {
                    if (tokenParts.length === 3) {
                        text = stringify(relatedGroup.properties)
                    } else {
                        const property = relatedGroup.properties
                            ? getPropertyValueByPath(relatedGroup.properties, tokenParts.slice(3))
                            : undefined
                        text = this.webhookEscape(property)
                    }
                }
            }
        } else {
            throw new Error()
        }
        return text
    }
}
