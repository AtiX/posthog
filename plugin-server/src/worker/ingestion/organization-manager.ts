import LRU from 'lru-cache'

import { RawOrganization, Team, TeamId } from '../../types'
import { DB } from '../../utils/db/db'
import { timeoutGuard } from '../../utils/db/utils'
import { TeamManager } from './team-manager'

const ONE_DAY = 24 * 60 * 60 * 1000

export class OrganizationManager {
    db: DB
    teamManager: TeamManager
    organizationCache: LRU<RawOrganization['id'], RawOrganization>
    availableFeaturesCache: LRU<TeamId, string[]>

    constructor(db: DB, teamManager: TeamManager) {
        this.db = db
        this.teamManager = teamManager
        this.organizationCache = new LRU({
            max: 10_000,
            // 30 seconds
            maxAge: 30_000,
            // being explicit about the fact that we want to update
            // the team cache every 2min, irrespective of the last access
            updateAgeOnGet: false,
        })
        this.availableFeaturesCache = new LRU({
            max: 10_000,
            maxAge: ONE_DAY,
            // being explicit about the fact that we want to update
            // the team cache every day, irrespective of the last access
            updateAgeOnGet: false,
        })
    }

    public async fetchOrganization(organizationId: RawOrganization['id']): Promise<RawOrganization | null> {
        const cachedOrganization = this.organizationCache.get(organizationId)
        if (cachedOrganization) {
            return cachedOrganization
        }

        const timeout = timeoutGuard(`Still running "fetchOrganization". Timeout warning after 30 sec!`)
        try {
            const organization: RawOrganization | null = (await this.db.fetchOrganization(organizationId)) || null
            if (organization) {
                this.organizationCache.set(organizationId, organization)
            }
            return organization
        } finally {
            clearTimeout(timeout)
        }
    }

    public async hasAvailableFeature(teamId: TeamId, feature: string, team?: Team): Promise<boolean> {
        const cachedAvailableFeatures = this.availableFeaturesCache.get(teamId)

        if (cachedAvailableFeatures !== undefined) {
            return cachedAvailableFeatures.includes(feature)
        }

        const _team = team || (await this.teamManager.fetchTeam(teamId))

        if (!_team) {
            return false
        }

        const organization = await this.fetchOrganization(_team.organization_id)
        const availableFeatures = organization?.available_features || []
        this.availableFeaturesCache.set(teamId, availableFeatures)

        return availableFeatures.includes(feature)
    }

    public resetAvailableFeatureCache(organizationId: string) {
        this.availableFeaturesCache.reset()
        this.organizationCache.del(organizationId)
    }
}
