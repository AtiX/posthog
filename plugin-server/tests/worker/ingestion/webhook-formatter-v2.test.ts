import { ISOTimestamp, PostIngestionEvent, Team } from '../../../src/types'
import { WebhookFormatterOptions, WebhookFormatterV2 } from '../../../src/worker/ingestion/webhook-formatter-v2'

type TestWebhookFormatterOptions = Partial<WebhookFormatterOptions> & {
    personProperties?: PostIngestionEvent['person_properties']
}

describe('WebhookFormatterV2', () => {
    const team = { id: 123, person_display_name_properties: null } as Team
    const event: PostIngestionEvent = {
        event: '$pageview',
        eventUuid: '123',
        teamId: 123,
        distinctId: 'WALL-E',
        person_properties: { email: 'test@posthog.com' },
        person_created_at: '2021-10-31T00%3A44%3A00.000Z' as ISOTimestamp,

        elementsList: [],
        properties: { $browser: 'Chrome' },
        timestamp: '2021-10-31T00%3A44%3A00.000Z' as ISOTimestamp,
        groups: {
            organization: {
                index: 1,
                type: 'organization',
                key: '123',
                properties: { name: 'PostHog', plan: 'paid' },
            },

            project: {
                index: 2,
                type: 'project',
                key: '234',
                properties: {},
            },
        },
    }

    const createFormatter = (options?: TestWebhookFormatterOptions) => {
        return new WebhookFormatterV2({
            sourceName: options?.sourceName ?? 'action1',
            sourcePath: options?.sourcePath ?? '/action/1',
            event: {
                ...(options?.event ?? event),
                person_properties: options?.personProperties ?? event.person_properties,
            },
            team: options?.team ?? team,
            siteUrl: options?.siteUrl ?? 'http://localhost:8000',
        })
    }

    beforeEach(() => {
        process.env.NODE_ENV = 'test'
    })

    describe('webhook formatting options', () => {
        const cases: [string, TestWebhookFormatterOptions][] = [
            ['{{person}}', {}],
            ['{{person.link}}', {}],
            ['{{user.name}}', {}], // Alias for person name
            ['{{user.browser}}', {}], // Otherwise just alias to event properties
            ['{{action.name}}', {}],
            ['{{action.name}} was done by {{user.name}}', {}],
            ['{{source.name}}', {}],
            ['{{source.name}} was done by {{user.name}}', {}],
            // Handle escaping brackets
            ['{{action.name\\}} got done by \\{{user.name\\}}', {}],
            ['{{event}}', {}],
            ['{{event.uuid}}', {}],
            ['{{event.name}}', {}], // Alias for event name
            ['{{event.event}}', {}],
            ['{{event.distinct_id}}', {}],
            [
                '{{person}}',
                {
                    personProperties: {
                        imię: 'Grzegorz',
                        nazwisko: 'Brzęczyszczykiewicz',
                    },
                    team: { ...team, person_display_name_properties: ['nazwisko'] },
                },
            ],
            [
                '{{person.properties.enjoys_broccoli_on_pizza}}',
                {
                    personProperties: { enjoys_broccoli_on_pizza: false },
                },
            ],
            [
                '{{person.properties.pizza_ingredient_ranking}}',
                {
                    personProperties: { pizza_ingredient_ranking: ['pineapple', 'broccoli', 'aubergine'] },
                },
            ],
            ['{{user.missing_property}}', {}],
            ['{{event}}', { event: { ...event, eventUuid: '**)', event: 'text](yes!), [new link' } }], // Special escaping
            [
                '{{user.name}} from {{user.browser}} on {{event.properties.page_title}} page with {{event.properties.fruit}}, {{event.properties.with space}}',
                {
                    event: {
                        ...event,
                        distinctId: '2',
                        properties: { $browser: 'Chrome', page_title: 'Pricing', 'with space': 'yes sir' },
                    },
                },
            ],
            ['{{groups}}', {}],
            ['{{groups.missing}}', {}],
            ['{{groups.organization}}', {}],
            ['{{groups.organization.properties.plan}}', {}],
            ['{{groups.project}}', {}], // No-name one

            ['{ "event_properties": {{event.properties}}, "person_link": "{{person.link}}" }', {}], // JSON object
        ]

        it.each(cases)('%s %s', (template, options) => {
            const formatter = createFormatter(options)
            const message = formatter.format(template)
            // For non-slack messages the text is always markdown
            expect(message).toMatchSnapshot()
        })
    })
})
