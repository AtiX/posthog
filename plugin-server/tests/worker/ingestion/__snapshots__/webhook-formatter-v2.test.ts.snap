// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WebhookFormatterV2 webhook formatting options { "event_properties": {{event.properties}}, "person_link": "{{person.link}}" } {} 1`] = `"{ \\"event_properties\\": {\\"$browser\\":\\"Chrome\\"}, \\"person_link\\": \\"http://localhost:8000/project/123/person/WALL-E\\" }"`;

exports[`WebhookFormatterV2 webhook formatting options {{action.name\\}} got done by \\{{user.name\\}} {} 1`] = `"{{action.name}} got done by {{user.name}}"`;

exports[`WebhookFormatterV2 webhook formatting options {{action.name}} {} 1`] = `"action1"`;

exports[`WebhookFormatterV2 webhook formatting options {{action.name}} was done by {{user.name}} {} 1`] = `"action1 was done by test@posthog.com"`;

exports[`WebhookFormatterV2 webhook formatting options {{event.distinct_id}} {} 1`] = `"WALL-E"`;

exports[`WebhookFormatterV2 webhook formatting options {{event.event}} {} 1`] = `"$pageview"`;

exports[`WebhookFormatterV2 webhook formatting options {{event.name}} {} 1`] = `"$pageview"`;

exports[`WebhookFormatterV2 webhook formatting options {{event.uuid}} {} 1`] = `"123"`;

exports[`WebhookFormatterV2 webhook formatting options {{event}} { event: [Object] } 1`] = `"{\\"uuid\\":\\"**)\\",\\"event\\":\\"text](yes!), [new link\\",\\"timestamp\\":\\"2021-10-31T00%3A44%3A00.000Z\\",\\"distinct_id\\":\\"WALL-E\\",\\"link\\":\\"http://localhost:8000/project/123/events/**)/2021-10-31T00%253A44%253A00.000Z\\",\\"properties\\":{\\"$browser\\":\\"Chrome\\"}}"`;

exports[`WebhookFormatterV2 webhook formatting options {{event}} {} 1`] = `"{\\"uuid\\":\\"123\\",\\"event\\":\\"$pageview\\",\\"timestamp\\":\\"2021-10-31T00%3A44%3A00.000Z\\",\\"distinct_id\\":\\"WALL-E\\",\\"link\\":\\"http://localhost:8000/project/123/events/123/2021-10-31T00%253A44%253A00.000Z\\",\\"properties\\":{\\"$browser\\":\\"Chrome\\"}}"`;

exports[`WebhookFormatterV2 webhook formatting options {{groups.missing}} {} 1`] = `"(event without group 'missing')"`;

exports[`WebhookFormatterV2 webhook formatting options {{groups.organization.properties.plan}} {} 1`] = `"paid"`;

exports[`WebhookFormatterV2 webhook formatting options {{groups.organization}} {} 1`] = `"{\\"index\\":1,\\"type\\":\\"organization\\",\\"key\\":\\"123\\",\\"properties\\":{\\"name\\":\\"PostHog\\",\\"plan\\":\\"paid\\"}}"`;

exports[`WebhookFormatterV2 webhook formatting options {{groups.project}} {} 1`] = `"{\\"index\\":2,\\"type\\":\\"project\\",\\"key\\":\\"234\\",\\"properties\\":{}}"`;

exports[`WebhookFormatterV2 webhook formatting options {{groups}} {} 1`] = `"{\\"organization\\":{\\"index\\":1,\\"type\\":\\"organization\\",\\"key\\":\\"123\\",\\"properties\\":{\\"name\\":\\"PostHog\\",\\"plan\\":\\"paid\\"}},\\"project\\":{\\"index\\":2,\\"type\\":\\"project\\",\\"key\\":\\"234\\",\\"properties\\":{}}}"`;

exports[`WebhookFormatterV2 webhook formatting options {{person.link}} {} 1`] = `"http://localhost:8000/project/123/person/WALL-E"`;

exports[`WebhookFormatterV2 webhook formatting options {{person.properties.enjoys_broccoli_on_pizza}} { personProperties: [Object] } 1`] = `"false"`;

exports[`WebhookFormatterV2 webhook formatting options {{person.properties.pizza_ingredient_ranking}} { personProperties: [Object] } 1`] = `"[\\"pineapple\\",\\"broccoli\\",\\"aubergine\\"\\\\]"`;

exports[`WebhookFormatterV2 webhook formatting options {{person}} { personProperties: [Object], team: [Object] } 1`] = `"{\\"properties\\":{\\"imię\\":\\"Grzegorz\\",\\"nazwisko\\":\\"Brzęczyszczykiewicz\\"},\\"link\\":\\"http://localhost:8000/project/123/person/WALL-E\\"}"`;

exports[`WebhookFormatterV2 webhook formatting options {{person}} {} 1`] = `"{\\"properties\\":{\\"email\\":\\"test@posthog.com\\"},\\"link\\":\\"http://localhost:8000/project/123/person/WALL-E\\"}"`;

exports[`WebhookFormatterV2 webhook formatting options {{source.name}} {} 1`] = `"action1"`;

exports[`WebhookFormatterV2 webhook formatting options {{source.name}} was done by {{user.name}} {} 1`] = `"action1 was done by test@posthog.com"`;

exports[`WebhookFormatterV2 webhook formatting options {{user.browser}} {} 1`] = `"Chrome"`;

exports[`WebhookFormatterV2 webhook formatting options {{user.missing_property}} {} 1`] = `"undefined"`;

exports[`WebhookFormatterV2 webhook formatting options {{user.name}} {} 1`] = `"test@posthog.com"`;

exports[`WebhookFormatterV2 webhook formatting options {{user.name}} from {{user.browser}} on {{event.properties.page_title}} page with {{event.properties.fruit}}, {{event.properties.with space}} { event: [Object] } 1`] = `"test@posthog.com from Chrome on Pricing page with undefined, yes sir"`;
